### DISCUSSION TITLE
Erasing SSD's and NVME's

### DISCUSSION THREAD

**Discussion by bigvictorio** [PartialVolume/shredos.x86_64#156] [Q&A] [answered]
Hello,

I was wondering which method Shredos uses, especially for Sata/M2 SSDs and NVME SSD.
In the documentation, there was a planned feature of adding **ATA SECURE ERASE** method, or **nvme-cli** library for NVME disks. 

Is Sherdos just filling all the blocks on the disks, so Shredos doesn't recognize the difference between disks? Or my understanding is incorrect.
If there is someone who could maybe explain to me how does deletion of NVMe or SATA SSDs work in this software, or provide more detailed information? I would be very thankful

Thanks :)

**Comment by PartialVolume** [comment, marked as answer]
ATA secure erase is scheduled for release 0.35 or 0.36. So at the moment it simply treats a SSD or NVME as if they were an ordinary spinning disc, ie just filling all the blocks. You can do a ATA secure erase  using hdparm or  nvme-cli by running the commands in the ALT F2 virtual terminal in ShredOS.

In the 0.35/0.36  release there will be a new method 'ATA Secure erase (HDD/SSD/NvMe)' built into the nwipe GUI that issues the ATA secure erase command and lets drive do the wipe. nwipe will estimate the progress on screen so it will look similar to wiping a HDD.

**Reply by uiob** [reply]
Is this still being worked on? Can we expect it in future releases?

**Reply by PartialVolume** [reply]
That's the plan. Unfortunately I'm really busy on things unrelated to ShredOS/Nwipe at the moment, so no timeline.

**Reply by SantosSi** [reply]
Can we expect this feature in the version after 0.38?

**Reply by DaanSelen** [reply]
> That's the plan. Unfortunately I'm really busy on things unrelated to ShredOS/Nwipe at the moment, so no timeline.

Is this in nwipe 0.39?

**Comment by bigvictorio** [comment]
Hello,

thanks for the response. Is there any expected release date for 0.35/0.36 or for nvme-cli/hdparm (or maybe month, year, etc)? Or it is just planned for the future?

**Reply by PartialVolume** [reply]
I'm currently finishing off the PDF report and will more than likely do a 0.35 release next week or the week after. I will then do a full release of ShredOS based on the 0.35 version of ShredOS, that will be the week after. ATA secure erase will be done immediately after that. I've given myself a couple of months, but may do it quicker.. but as soon as ATA secure erase is complete I would release as 0.36. So probably two months between the two releases. So i guess I'm looking at sometime in September .. fingers crossed for ATA secure erase... famous last words :-)

**Reply by PartialVolume** [reply]
Sorry October, not September. :-)

**Comment by bigvictorio** [comment]
Okay thanks, that's a pity but I will be looking forward to it ⏩ 

> In the 0.35/0.36 release there will be a new method 'ATA Secure erase (HDD/SSD/NvMe)' built into the nwipe GUI that issues the ATA secure erase command and lets drive do the wipe.

Are you thinking, that it will be just an option in GUI to select the deletion method? Or some automation, because it would be beautiful to just connect all disks to the computer running shredos, the system will recognize them and delete them accordingly by their type. (HDD, SATA SSD, NVME)

I'm thinking something like this script for Parted can maybe be an inspiration: https://github.com/tslight/disknukem

**Reply by PartialVolume** [reply]
>the system will recognize them and delete them accordingly by their type. (HDD, SATA SSD, NVME).

Yes, that's certainly a possibility, i.e in config there could be an option **Always wipe SSD/NvMe with ATA secure erase** If that is set then only connected HDD's will be wiped with the selected method, while the SSD/NvMe drives will be wiped with secure erase.

**Reply by PartialVolume** [reply]
>I'm thinking something like this script for Parted can maybe be an inspiration: https://github.com/tslight/disknukem

Thanks for that link. It's always useful to see how others do it. I could always include those scripts in shredos if anybody is interested.

**Reply by bigvictorio** [reply]
>  I could always include those scripts in ShredOs if anybody is interested.

Maybe this could serve as a temporary solution until you develop something robust and you will be satisfied with it.
It's also better for people who would like to properly and more securely erase their disks.
And with the ISC license assigned to Disknukem, I don't think there should be some problem. 
Maybe some bugs I found:
- **disknukem** is not "deleting" disk simultaneously (I mean at least in the GUI")
-  there are some old libraries that at least Parted stopped using. (I think, to make that script work I had to replace D3udd with dd)
-  there is only one iteration

But it would be very useful

**Comment by henk717** [comment]
Did this get added? I can't find it in the UI
Also worth pointing out, drives can lack this feature or block it. In those cases a drive wide TRIM is helpful and often good enough / quick.

**Reply by PartialVolume** [reply]
>Did this get added? I can't find it in the UI.

No it's not been added yet. There has been too many other things going on in my life. However it's still on my list, but I can't give a timeline as I just don't know at the moment. I'm focused on fixing some issues in nwipe, which I've done then I'm releasing a new release of nwipe and an updated ShredOS with the latest kernel & drivers. Then back to adding more features to nwipe.

**Reply by Darth-NullVortex** [reply]
I truly appreciate your dedication to this project. It's brilliant! I understand the challenges of balancing priorities and time constraints. Your plan to integrate NVMe and SATA erase commands into the UI sounds excellent. This addition would be incredibly valuable, especially with the PDF report. Whenever you're able to implement it, I'm sure it'll be worth the wait. Thank you for considering users' needs and continuously improving the software. Keep up the great work! I'm curious, is there any way for non-developers like myself to support this effort?

**Comment by brazier** [comment]
I came here looking for this, but the method of using ALT+F2 did the job. could it be an idea to add the hdparm and nvme method to the wiki for now, and/or add a shell script so that for now you could do something like erase-nvme/erase-ssd /dev/device. (i can help out with both if needed)

**Reply by PartialVolume** [reply]
Adding the command line procedures for secure erasing using nvme or hdparm to the wiki or README.md would be useful. If you would like to do that it would be appreciated.

A pull request is preferred however if your not familiar with that, then a post containing the details and I can transfer it to the readme.

**Reply by Firminator** [reply]
Brazier, in case you missed it... the instructions have been added recently
@ https://github.com/PartialVolume/shredos.x86_64#wipe-ssd-and-nvme-using-hdparm-and-nvme-cli

**Reply by brazier** [reply]
@Firminator Thanks, I know, since i was the one that added it :)

**Reply by Firminator** [reply]
u the man!  Important contribution. Sorry for missing the correlation.

**Reply by Darth-NullVortex** [reply]
@PartialVolume Hi, I just wanted to share an important observation that might be useful for your implementation of the secure erase feature. It may be necessary to wipe the partition table and metadata at the beginning of the drive, as I recently noticed that an older SSD I wiped using hdparm still retained its partition table. Since the sanitize function is not supported for this drive, a wipe remains necessary. Additionally, even when sanitize is supported, it does not necessarily remove the partition table. I hope this information helps with your development process.

**Reply by PartialVolume** [reply]
Thanks for the info. Can you name and shame the make and model of the drive that failed to wipe itself using it's own firmware?

That looks like a good example of where a secure sanitize (using the drives firmware) should be followed by a standard traditional prng wipe, followed by a prng verification rolled into a single method within nwipe which would have resulted in wiping the drive fully and verifying it was correctly wiped, or at least as far as it's possible with an SSD. [Edited]

**Reply by Darth-NullVortex** [reply]
Certainly! It’s the Samsung SSD 860 EVO 250GB with firmware RVT03B6Q. The drive appears to have been replaced, as these laptops typically house older models as well as completely different ones, so I am rather surprised to find it’s not actually that outdated.

**Reply by Firminator** [reply]
There might be a slight chance that Samsung fixed the drive's firmware Secure Erase/Sanitize implementation in their latest update @ https://semiconductor.samsung.com/resources/software-resources/Samsung_SSD_860_EVO_RVT0**4B**6Q_Win.iso
Your drive is one version behind. But without a changelog this is like shooting in the dark and just hoping that they fixed it. Might be worth a try though.

**Comment by dmf-hd** [comment]
New to the subject I have a couple questions using shredos‘ secure erase/sanitize functions for ATA M.2 SSDs and NVME M.2s SSDs. 

I am working in a hospital/university environment where the aspect of wiping data completely is crucial. We are planning to replace older laptops that are still good and give them to employees and their families. To meet the wiping criteria of the German Agency for Information Security we need to make sure no data from former use can be recovered. The older laptops do not have the secure erase function in their bios. So I used a newer laptop and put the SSDs of the older laptops in there and erased them with that bios secure erase function. To cross check if the erase procedure was 100% successful (it was per the report) I put the SSDs in an external housing and connected it to a PC thru USB. I then used Disk Drills Deep Scan function to check if there was anything on that disk that could be recovered. When the scan was over Disk Drill had found a system volume information folder with around 80 kb of data that could be recovered.
Now my question is if that is data that was not wiped from the SSD in the first place or that the data was written on the disk as a consequence of connecting the SSD to the PC thru USB?
And would the precence of that data be negligible in the context of that wiping process? What say you?

Now this is where ShreOs comes in. There is still no way to use wiping processes for SSDs thru the GUI but thru commands in the terminal. So if the SSDs would be placed in their respective laptops and wiped in that fashion would they now be completely empty? And what would happen if I took them out and connected them to a PC again to run a Data Recovery Scan. Would it be empty or would I find the system volume information folder again because it’s applied thru the Windows routine when connecting thru USB?

Another question, what is your recommended (most secure) way of erasing data from HDDs using ShredOs. Time is not so much a criteria. But it should avoid unnecessary action and on the other hand should not stress the hard drive more than needed.
Thank you for answering.

**Comment by eschoeller** [comment]
Hello! I am hoping these set of features are still on your roadmap. I have been extremely pleased with ShredOS and the autonuke feature so far. We have built a custom ShredOS/autonuke ISO and integrated that into our ansible automation to wipe an entire cluster of Dell servers before installing OpenShift / ODF onto them. What we are starting to find is that ODF has trouble installing itself since there is leftover metadata on the NVMe drives from previous ODF installs, even after autonuke/nwipe has run on them. It would be fantastic for us if these secure erase options were integrated into ShredOS and available as autonuke style command arguments so our entire automation workflow can run smoothly. 
Furthermore, as we surplus hardware we plan to use this same custom ShredOS image to automatically wipe hardware before putting it up for auction and it would be great to know that all the metadata on SSDs and NVMe drives has been cleared in addition to whichever nwipe options we have set. 
Thanks for all your hard work and dedication to this project, it's really great!

**Reply by PartialVolume** [reply]
Thanks, I'm glad you find it useful.

I'm interested in this left over metadata. Can you provide more detail? If you are verifying the disc it should have failed indicating some issue with the drive.

I've come across a couple of SSDs that write to the wrong locations intermittently due to a fault but don't return write errors, a fault with the drive only detectable if a verification pass is included in your wipe method.

**Reply by DaanSelen** [reply]
> Thanks, I'm glad you find it useful.
> 
> I'm interested in this left over metadata. Can you provide more detail? If you are verifying the disc it should have failed indicating some issue with the drive.
> 
> I've come across a couple of SSDs that write to the wrong locations intermittently due to a fault but don't return write errors, a fault with the drive only detectable if a verification pass is included in your wipe method.

Ah so as long as we have a verify pass. It should be ok.

**Reply by PartialVolume** [reply]
Just to add, a verification on a random pass is preferable to a verification on a zeros or ones as it's much more likely to discover stuck bits in a data channel which you may miss if you only do a zeros and verify or ones and verify, especially if it's a single stuck bit. Just gives you that extra bit of certainty that your hardware is functioning correctly.

**Reply by DaanSelen** [reply]
> Just to add, a verification on a random pass is preferable to a verification on a zeros or ones as it's much more likely to discover stuck bits in a data channel which you may miss if you only do a zeros and verify or ones and verify, especially if it's a single stuck bit. Just gives you that extra bit of certainty that your hardware is functioning correctly.

Can you quickly verify that currently ShredOS with nwipe 0.38 or 0.39 does clean SSD's correctly?

**Reply by PartialVolume** [reply]
> > Just to add, a verification on a random pass is preferable to a verification on a zeros or ones as it's much more likely to discover stuck bits in a data channel which you may miss if you only do a zeros and verify or ones and verify, especially if it's a single stuck bit. Just gives you that extra bit of certainty that your hardware is functioning correctly.
> 
> Can you quickly verify that currently ShredOS with nwipe 0.38 or 0.39 does clean SSD's correctly?

Also answered elsewhere but yes, you can wipe a SSD correctly but only if you use the hdparm or nvme command line tools included in ShredOS to issue a secure erase command to the drive. Nwipe alone is not sufficient to completely erase areas of the disk not accessible by block access. A PRNG + verification using nwipe plus an ATA secure erase using hdparm or nvme is preferable to only using nwipe until such time that nwipe includes a ATA secure erase option.

**Reply by DaanSelen** [reply]
https://www.usenix.org/legacy/events/fast11/tech/full_papers/Wei.pdf

Is this something we can cite? I would love your comment on this.

**Reply by PartialVolume** [reply]
It's an excellent article that's been discussed before in the nwipe/ShredOS issues and discussions several times. It's why we have a detailed section on wiping NVME and SSD in the ShredOS readme.

It's also why I prefer adding a random wipe in addition to a secure erase. Random because random data is highly uncompressible, which is relevant to drives that use compression techniques as discussed in the article and in addition a random traditional wipe of one to three rounds because, as again discussed in the article, there's no ATA commands to verify that all data in the drive has been removed by the secure erase command. You have to trust the ATA secure erase command did what it was supposed to do and erased areas of memory that we don't have access to, unless you dismantle the drive and put the chips in a reader.

So to cover all possible bases, including drive firmware that isn't wiping the drive as effectively as it should, a 1-3 round PRNG (random) with verification followed by the enhanced secure erase. I can't over emphasize the importance of uncompressible random data that can be verified. By its very nature of being uncompressible (try compressing random data generated by nwipe, you always end up with a bigger file than when you started) from a statistical point of view much of the over provisioned RAM will be written. This is just my point of view but writing zeros or ones to a SSD or nvme that uses compression techniques is a complete waste of time if the objective is to erase all memory, as it may appear the drive is empty but it is in fact not, depending on the make/model/firmware of the drive. Hence why you should always use random data as the drive has to write those TBytes of data uncompressed to memory and then has to reproduce for verification without using compression techniques to minimise the memory in use.

**Reply by DaanSelen** [reply]
> It's an excellent article that's been discussed before in the nwipe/ShredOS issues and discussions several times. It's why we have a detailed section on wiping NVME and SSD in the ShredOS readme.
> 
> It's also why I prefer adding a random wipe in addition to a secure erase. Random because random data is highly uncompressible, which is relevant to drives that use compression techniques as discussed in the article and in addition a random traditional wipe of one to three rounds because, as again discussed in the article, there's no ATA commands to verify that all data in the drive has been removed by the secure erase command. You have to trust the ATA secure erase command did what it was supposed to do and erased areas of memory that we don't have access to, unless you dismantle the drive and put the chips in a reader.
> 
> So to cover all possible bases, including drive firmware that isn't wiping the drive as effectively as it should, a 1-3 round PRNG (random) with verification followed by the enhanced secure erase. I can't over emphasize the importance of uncompressible random data that can be verified. By its very nature of being uncompressible (try compressing random data generated by nwipe, you always end up with a bigger file than when you started) from a statistical point of view much of the over provisioned RAM will be written. This is just my point of view but writing zeros or ones to a SSD or nvme that uses compression techniques is a complete waste of time if the objective is to erase all memory, as it may appear the drive is empty but it is in fact not, depending on the make/model/firmware of the drive. Hence why you should always use random data as the drive has to write those TBytes of data uncompressed to memory and then has to reproduce for verification without using compression techniques to minimise the memory in use.

If possible I would love to discuss these details with you in some kind of messaging, email - DM - whatever, I am afraid that I somehow got the wrong matrix address. But If you are open to that?

**Reply by DaanSelen** [reply]
And also how does this compare: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-88r2.pdf

**Reply by eschoeller** [reply]
Thanks for all your feedback on this. I think we also simultaneously hit a bug with ceph/ODF/dmcrypt/OpenShift in the process of all this - and so it's possible the "incomplete" wiping may not have actually been the culprit, and if so I certainly apologize for jumping to that conclusion. But this has been a tricky one to pin down. I see your points about the importance of random data and verification - and I think I'll switch my autonuke image to use those. My main goal was to run through this operation quickly so we could do clean re-installs on the hardware, not necessarily concerned about releasing the drives out into the wild. 
I may not have another opportunity to test these scenarios again on this hardware since it's being forced into production but if I do get another chance I will certainly try to fetch as much metadata from the drives *after* I run an autonuke with random data and verification to see if ODF complains again about the drive not being "fully wiped". 
For reference this is what the RedHat recommendations are for fully wiping disks so they are ready for ODF:
https://access.redhat.com/solutions/7115651

if you don't have access to that, we ended up running "Method 2" which is:

```
DISK="/dev/sdX"
DISK="/dev/nvme1n1"

# Wipe the partition table off the disk to a fresh, usable state.
wipefs -fa $DISK

# Wipe certain areas of the disk to remove Ceph Metadata which may be present.
for gb in 0 1 10 100 1000; do dd if=/dev/zero of="$DISK" bs=1K count=200 oflag=direct,dsync seek=$((gb * 1024**2)); done

# This might not be supported on all devices.
blkdiscard $DISK
```

Method 1 is:
```
Syntax: # /usr/bin/podman run --authfile /var/lib/kubelet/config.json --rm -ti --privileged --device <device-path> --entrypoint ceph-bluestore-tool <rook-ceph-operator-image> zap-device --dev <device-path> --yes-i-really-really-mean-it

Examples:
# /usr/bin/podman run --authfile /var/lib/kubelet/config.json --rm -ti --privileged --device /dev/sdc --entrypoint ceph-bluestore-tool registry.redhat.io/odf4/rook-ceph-rhel9-operator@sha256:60623084100a785f7eaf4bd3a4f87de4982daa37e30849bdb0b8304d3d1b7cd5 zap-device --dev /dev/sdc --yes-i-really-really-mean-it

# /usr/bin/podman run --authfile /var/lib/kubelet/config.json --rm -ti --privileged --device /dev/nvme1n1 --entrypoint ceph-bluestore-tool registry.redhat.io/odf4/rook-ceph-rhel9-operator@sha256:60623084100a785f7eaf4bd3a4f87de4982daa37e30849bdb0b8304d3d1b7cd5 zap-device --dev /dev/nvme1n1 --yes-i-really-really-mean-it
```

**Comment by brazier** [comment]
Back when i wrote the instructions in the readme, i ended up writing a script, to automaticly detect disks and their capabilities, then wipe with what is available. Depends on nvme-cli, hdparm, od & pv(dont remember if od/pv are available in shredos but can easily be modified to use dd/hexdump). Could perhaps be added to shredos while waiting for the final product.

https://github.com/brazier/wiper
or the way i use it:
`curl -sL https://wiper.004269.xyz | bash -s -- [options]`

Edit: Also noticed there is a missing "#" in the link to "wipe nvme", probably has been there since i wrote it.  do you wanna fix or do you want a PR just for that :P
```
   1. [Wipe SSD](#wipe-ssd)
   1. [Wipe NVME](wipe-nvme)
```

**Reply by PartialVolume** [reply]
I can fix that. I'll look at including this script on the next release.

**Reply by PartialVolume** [reply]
> Edit: Also noticed there is a missing "#" in the link to "wipe nvme", probably has been there since i wrote it. do you wanna fix or do you want a PR just for that :P

committed https://github.com/PartialVolume/shredos.x86_64/commit/4e544986c1bac2d7b46a8e093b2b9b96a42de4d3
